src/java/org/apache/cassandra/io/compress/AdaptiveCompressor.java

```diff
@@ -0,0 +1,390 @@
+                currentCompressionLevel--;
+            // If we're not under heavy write pressure, or we're spending very little time compressing data,
+            // we can increase the compression level and get some space savings at a low performance overhead:
+            else if (compressionLevelTarget > currentCompressionLevel || avgUtilisation < 0.05)
+                currentCompressionLevel++;
+
+            currentCompressionLevel = clampCompressionLevel(currentCompressionLevel);
+            return currentCompressionLevel;
+        }
+
+        /**
+         * Must be called after compressing a chunk,
+         * so we can measure how much time we spend in compressing vs time spent not-compressing.
+         */
+        public void recordCompressionDuration(long startTime, long endTime)
+        {
+            this.lastCompressionDuration = endTime - startTime;
+            this.lastCompressionStartTime = startTime;
+        }
+
+    }
+
+
+    private AdaptiveCompressor(Params params)
+    {
+        this.params = params;
+        this.state = new ThreadLocal<>();
+    }
+    
+    @Override
+    public int initialCompressedBufferLength(int chunkLength)
+    {
+        return (int) Zstd.compressBound(chunkLength);
+    }
+
+
+    @Override
+    public void compress(ByteBuffer input, ByteBuffer output) throws IOException
+    {
+        try
+        {
+            State state = getThreadLocalState();
+            int compressionLevel = state.adjustAndGetCompressionLevel();
+            long startTime = System.nanoTime();
+            Zstd.compress(output, input, compressionLevel, true);
+            long endTime = System.nanoTime();
+            state.recordCompressionDuration(startTime, endTime);
+            metrics.get(params.uses).updateFrom(state);
+        }
+        catch (Exception e)
+        {
+            throw new IOException("Compression failed", e);
+        }
+
+    }
+
+    @Override
+    public int uncompress(byte[] input, int inputOffset, int inputLength, byte[] output, int outputOffset) throws IOException
+    {
+        long dsz = Zstd.decompressByteArray(output, outputOffset, output.length - outputOffset,
+                                            input, inputOffset, inputLength);
+
+        if (Zstd.isError(dsz))
+            throw new IOException(String.format("Decompression failed due to %s", Zstd.getErrorName(dsz)));
+
+        return (int) dsz;
+    }
+
+    @Override
+    public void uncompress(ByteBuffer input, ByteBuffer output) throws IOException
+    {
+        try
+        {
+            Zstd.decompress(output, input);
+        } catch (Exception e)
+        {
+            throw new IOException("Decompression failed", e);
+        }
+    }
+
+    @Override
+    public BufferType preferredBufferType()
+    {
+        return BufferType.OFF_HEAP;
+    }
+
+    @Override
+    public Set<Uses> recommendedUses()
+    {
+        return Set.of(params.uses);
```

Comment:
This can use `EnumSet.of`.

Further, IMHO it makes better sense to return both here; or rather, add a new method to adapt the compressor for a certain use, e.g.
```
default ICompressor forUse(Uses use)
{
    return recommendedUses.contains(use) ? this : null;
}
``` 
in `ICompressor`, overridden to change params and recreate in `AdaptiveCompressor`.

Reply:
Actually, such a change is also necessary to preserve encryption for HCD.

Reply:
+1 to `EnumSet.of`

However, I don't agree with returning both uses here. 
This is because a particular `AdaptiveCompressor` can be either adapting to "flush pressure" or "compaction pressure" but not both. I don't want to accidentally end up using `AdaptiveCompressor` configured for compaction (aka GENERAL use) in flush, so it cannot advertise as FAST_COMPRESSION unless configured for flushing. 
The use is locked at the moment of creating the compressor. It could be actually two separate compressor classes, each for different use, but because they share like 99% of logic and the only real difference is defaults + pressure source, there is likely no point in separating them.

Reply:
But on the second thought, indeed, if someone configures flushCompression as `table` and then `AdaptiveCompression` is selected in the table schema... this would result in using a wrong source of pressure. So adding this `forUse` API looks like a good idea to me.

Reply:
This is precisely why I described an adaptation method: we do want to use a compressor derived from the current one, suitable for the flush usage. This can be done without requiring any changes to existing `ICompressor` implementations via the default method above, and is important for this patch, but also critical for preserving encryption when DSE moves to this code base.